packages/lesswrong/server/autocompleteEndpoint.ts

```diff
@@ -0,0 +1,428 @@
+import type { Express } from "express";
+import { getUserFromReq } from "./vulcan-lib/apollo-server/context";
+import { userIsAdmin } from "@/lib/vulcan-users/permissions";
+import { getAnthropicPromptCachingClientOrThrow } from "./languageModels/anthropicClient";
+import express from "express";
+import { PromptCachingBetaMessageParam } from "@anthropic-ai/sdk/resources/beta/prompt-caching/messages";
+import { Posts } from "@/lib/collections/posts";
+import { Comments } from "@/lib/collections/comments";
+import Revisions from "@/lib/collections/revisions/collection";
+import { turndownService } from "./editor/conversionUtils";
+import { formatRelative } from "@/lib/utils/timeFormat";
+import Users from "@/lib/vulcan-users";
+const { Readable } = require('stream');
+import { pipeline } from 'stream/promises'
+import { hyperbolicApiKey } from "@/lib/instanceSettings";
+
+
+
+const getParentComments = async (comment: DbComment) => {
+  const parentComments: DbComment[] = [];
+  let currentComment: DbComment | null = comment;
+  while (currentComment.parentCommentId && currentComment.parentCommentId !== currentComment._id) {
+    currentComment = await Comments.findOne({ _id: currentComment.parentCommentId });
+    if (currentComment) {
+      parentComments.push(currentComment);
+    } else {
+      break;
+    }
+  }
+
+  return parentComments.reverse();
+}
```

Comment:
This should probably be a sql query with a recursive CTE, like `parent_comments` in `PostsRepo.getPostsAndCommentsFromSubscriptions`.  (Can be a TODO if you don't want to mess with it now.  But I do expect the performance difference to be significant.)

Reply:
Eh, unless you are in deep comment threads this is only like 4 levels deep and not worth the complexity IMO